Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs

```diff
@@ -0,0 +1,210 @@
+
+        private void ReactWithTileAndEntities()
+        {
+            if (!Owner.TryGetComponent(out SolutionContainerComponent contents))
+                return;
+
+            var mapGrid = _mapManager.GetGrid(Owner.Transform.GridID);
+            var tile = mapGrid.GetTileRef(Owner.Transform.Coordinates.ToVector2i(Owner.EntityManager, _mapManager));
+
+            var solutionFraction = 1 / Math.Floor(_averageLifetime / ExposeDelay);
+
+            // Reagents react with the tile under the smoke
+            foreach (var reagentQuantity in contents.ReagentList.ToArray())
+            {
+                if (reagentQuantity.Quantity == ReagentUnit.Zero) continue;
+                var reagent = _prototypeManager.Index<ReagentPrototype>(reagentQuantity.ReagentId);
+                reagent.ReactionTile(tile, reagentQuantity.Quantity * solutionFraction);
+            }
+
+            foreach (var entity in tile.GetEntitiesInTileFast())
+            {
+                // Reagents touch all entities on the tile
+                foreach (var reagentQuantity in contents.ReagentList.ToArray())
+                {
+                    if (reagentQuantity.Quantity == ReagentUnit.Zero) continue;
+                    var reagent = _prototypeManager.Index<ReagentPrototype>(reagentQuantity.ReagentId);
+                    reagent.ReactionEntity(entity, ReactionMethod.Touch, reagentQuantity.Quantity * solutionFraction);
+                }
+
+                // Reagents enter the bloodstream of all entities without internals.
+                if (!entity.TryGetComponent(out BloodstreamComponent bloodstream))
+                    continue;
+
+                if (entity.TryGetComponent(out InternalsComponent internals) &&
+                    internals.BreathToolEntity != null &&
+                    internals.GasTankEntity != null &&
+                    internals.BreathToolEntity.TryGetComponent(out BreathToolComponent breathTool) &&
+                    breathTool.IsFunctional &&
+                    internals.GasTankEntity.TryGetComponent(out GasTankComponent gasTank) &&
+                    gasTank.Air != null)
+                    continue;
+
+                var cloneSolution = contents.Solution.Clone();
+                var transferAmount = ReagentUnit.Min(cloneSolution.TotalVolume * solutionFraction, bloodstream.EmptyVolume);
+                var transferSolution = cloneSolution.SplitSolution(transferAmount);
+
+                foreach (var reagentQuantity in transferSolution.Contents.ToArray())
+                {
+                    if (reagentQuantity.Quantity == ReagentUnit.Zero) continue;
+                    var reagent = _prototypeManager.Index<ReagentPrototype>(reagentQuantity.ReagentId);
+                    transferSolution.RemoveReagent(reagentQuantity.ReagentId,reagent.ReactionEntity(entity, ReactionMethod.Ingestion, reagentQuantity.Quantity));
+                }
+
+                bloodstream.TryTransferSolution(transferSolution);
```

Comment:
With no cloning you'd need to add it back into the smoke if the transfer failed.